### PR TITLE
Inventory-based SSH proxy/bastion host support, ssh_args + docs

### PR DESCRIPTION
**Pull request by nitzmahone**
Supersedes #9122 

First-class inventory var support for proxy/bastion/jump hosts seems to be increasingly requested. I've tried to make it reasonably comprehensive and well-documented, though I intentionally omitted proxy_password support (had some trouble stacking the file descriptors for the two-level-deep sshpass).

(CC @sivel)

### CODE REVIEW

**Comment by jimsmith** [comment]
+1 for this, it's a request that I rarely get asked but it's becoming more frequently.

**Comment by mpdehaan** [comment]
So this has been discussed a bit recently on the devel list, and I believe there may be some competing implementations, can you stop by ansible-devel to discuss?

Thanks!

**Comment by tzz** [comment]
I've needed this so definitely happy to see an implementation!

**Comment by quinot** [comment]
I'm just starting to use Ansible and was wondering whether this feature existed, or how to implement it if not. Would love to have this merged!

**Comment by rmetzler** [comment]
I really want this feature in Ansible, but after looking at the code, I would think you shouldn't add that many arguments to `connect`. Maybe use `*args` and `**kwargs`.

**Comment by nitzmahone** [comment]
Funny, that was my first thought when I looked at it too- "why are they passing all these args positionally?" I looked at the history, and the last couple of args were added later. I assumed there was some reason to do with the abstraction of the Connect stuff, and went for the "when in Rome" approach, but if reviewers feel the same way, I'm happy to convert it to use named args instead. It _is_ much cleaner, IMHO.

**Comment by rmetzler** [comment]
I want to say, I just said this as a user who needs this feature and searched through the different threads / PRs. I really don't have any authority regarding Ansible.

**Comment by igoratencompass** [comment]
I've been waiting for this kind of module definitely needed for dynamic aws inventory of mixed public and private instances! Can we please have this released asap???

**Comment by christophgysin** [comment]
I'm also looking forward to this!

File: lib/ansible/runner/connection_plugins/ssh.py [hunk 1]
```diff
@@ -58,10 +63,35 @@ def connect(self):
...
+            proxy_args += ["-o", "KbdInteractiveAuthentication=no",
+                           "-o", "PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey",
+                           "-o", "PasswordAuthentication=no"
+            ]
+
+            if self.proxy_user is not None and self.proxy_user != pwd.getpwuid(os.geteuid())[0]:
+                proxy_args += ["-o", "User="+self.proxy_user]
+            proxy_args += ["-o", "ConnectTimeout=%d" % self.runner.timeout]
+
+            self.common_args += ["-o", "ProxyCommand=ssh {proxy_host} {proxy_args} -W %h:%p".format(proxy_host=self.proxy_host, proxy_args=" ".join(proxy_args))]
```
Comment (christophgysin) [inline lib/ansible/runner/connection_plugins/ssh.py:92]: ssh -W asumes openssh-5.4 or later. Older versions of openssh need netcat to achieve this.

Comment (nitzmahone) [inline lib/ansible/runner/connection_plugins/ssh.py:92]: Yeah, I considered using netcat as a lowest-common-denominator approach. However, I saw some complaints around the openssh community about instability with various versions of netcat that the -W mode attempts to address (don't have a reference to share off the top of my head- it was several months ago). If a bastion host can't be upgraded to a reasonably recent version of openssh, the ssh_args escape hatch is still there- I think that beats the alternatives of:
- forcing everyone to use a potentially unstable approach that has another dependency (netcat)
- adding yet another mode-changing variable (ansible_ssh_proxy_mode=netcat or some such blah)
- version-sniffing ssh (rife with issues)

**Comment by anthcourtney** [comment]
Is there a timeline for this feature progressing / being integrated?

**Comment by rmetzler** [comment]
For everyone needing a workaround fast, this is a great tutorial: http://alexbilbie.com/2014/07/using-ansible-with-a-bastion-host/

But I also would prefer first class support in Ansible.

**Comment by mpeters** [comment]
That approach only works when you have a single bastion server for all your environments. This won't work if you have a separate bastion for say production, staging and integration.

**Comment by bcoca** [comment]
@mpeters it does, you just need to specify the diff bastions per host or host pattern (this is where naming conventions can help a lot).

**Comment by mpeters** [comment]
That's assuming you are naming your servers. How would you handle just IP addresses (since these are livestock, not pets) in different VPCs? The IPs can be repeated for a different environment. 

Also, how would you hook this up with Tower? (You'd need multiple Tower installs right?) Multiple Tower installs? 

What about a single environment that uses multiple VPCs for PCI, etc?

**Comment by bcoca** [comment]
no need for multiple ansible/tower installs, one thing you can do is set ssh args per group/host/inventory, passing specific ssh config for each.

In any case, this is a workaround, the patch is a good step into a more integrated solution.

**Comment by mpeters** [comment]
Can you really set ansible_ssh_args per group/host/inventory (that would be great)? The documentation doesn't list it with the rest of the ansible_ssh_\* args (http://docs.ansible.com/intro_inventory.html#list-of-behavioral-inventory-parameters) and the only place I see you can set it is in the ansible config file, which is global.

**Comment by rmetzler** [comment]
I had a quick look into the source and it seems like `ssh_args` is respected in 1.8.2. Not sure since when, though.

https://github.com/ansible/ansible/blob/v1.8.2/lib/ansible/constants.py#L165
https://github.com/ansible/ansible/blob/v1.8.2/lib/ansible/runner/connection_plugins/ssh.py#L55-L64

**Comment by mpeters** [comment]
But that just shows that it's configurable globally from the ansible configuration file. Not that it's possible to specify per group/host/inventory, right?

**Comment by snackycracky** [comment]
I also need the 'Ansible EC2 external inventory script' to detect the internal hosts. http://docs.ansible.com/intro_dynamic_inventory.html#example-aws-ec2-external-inventory-script
it only finds the jump-hosts for each environment but not the internal servers.

**Comment by dopry** [comment]
@bcoca,  Are you comfortable with this approach to bastion host support, or is there a competing implementation you prefer? I'm currently working on a VPC deployment and can put some cycles in testing and getting the PR up to date if you guys are interested in getting it merged.

**Comment by eriweb** [comment]
Thanks @nitzmahone, I'm testing this, but I seem to have trouble getting Ansible to wait for the proxycommand to finish.
I get this error message blazingly fast: "fatal: [r-4-VM] => SSH Error: data could not be sent to the remote host. Make sure this host can be reached over ssh"
but if I try the same SSH command manually it works just fine..

**Comment by bcoca** [comment]
I was thinking about integrating this soon now that v2 is almost 'there', but needs massive rebase, should probably wait till v2 is in place in devel

**Comment by bcoca** [comment]
the other thing I wanted to do is move away from having ansiible_ssh and go for ansible_connection (or at least conn) as the variable naming, so we can actually use the same vars for other protocols w/o asking 'why does winrm use ansible_ssh_user?'

**Comment by nitzmahone** [comment]
Yeah, this will probably just end up being a whole new PR for v2 (adding support for proxy chaining, new arg naming, etc), but now that I know cchurch's trick for smuggling vars in v1, I might backport the chaining stuff to this one as a standalone connection plugin. Let's leave this one open for now, and I'll try to prioritize the v2 work and backport.

**Comment by sivel** [comment]
This PR was superseded by the already merged https://github.com/ansible/ansible/pull/11908

**Pull request closed without merging**